src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -532,6 +544,142 @@ public static int GetCurrentProcessorId()
+        //   is in start-up mode. That would just add to the "rush hour" traffic.
+        private static int sCalibrationToDo;
+        private static int sCalibrationDone;
+        // 25 is chosen to budget the sampling under 100 msec total, assuming 4 msec per sample.
+        private const int CalibrationSamples = 25;
+        private static double[] sCalibrationSamples = new double[CalibrationSamples * 2];
+
+        private static void CalibrateOnce()
+        {
+            if (sCalibrationToDo >= CalibrationSamples)
+                return;
+
+            int sample = Interlocked.Increment(ref sCalibrationToDo) - 1;
+            if (sample >= CalibrationSamples)
+                return;
+
+            // Actual calibration step. Let's try to fit into ~4 msec.
+            double[] calibrationState = sCalibrationSamples;
+
+            int id = 0;
+            long t1 = 0;
+            long _1msec = Stopwatch.Frequency / 1000;
+            int iters = 1;
+
+            // double the sample size until it is 1 msec.
+            // we may spend up to 3 msec in this loop in a worst case.
+            while (t1 < _1msec)
+            {
+                iters *= 2;
+                t1 = Stopwatch.GetTimestamp();
+                for (int i = 0; i < iters; i++)
+                {
+                    id = GetCurrentProcessorNumber();
+                }
+                t1 = Stopwatch.GetTimestamp() - t1;
+            }
+
+            // assuming TLS cannot be a lot slower than ID, this should take 1 msec
+            long t2 = Stopwatch.GetTimestamp();
+            for (int i = 0; i < iters; i++)
+            {
+                UninlinedThreadStatic();
+            }
+            long t3 = Stopwatch.GetTimestamp();
+
+            // if we have useful measurements, record a sample
+            if (id >= 0 && t1 > 0 && t3 - t2 > 0)
+            {
+                calibrationState[sample * 2] = (double)t1 / iters;            // ID
+                calibrationState[sample * 2 + 1] = (double)(t3 - t2) / iters; // TLS
+            }
+            else
+            {
+                // API is not functional or clock did not go forward.
+                // just pretend it was a very expensive sample with default ratio.
+                calibrationState[sample * 2] = (double)Stopwatch.Frequency * 50; // 50 sec;
+                calibrationState[sample * 2 + 1] = Stopwatch.Frequency; // 1 sec
+            }
+
+            // If this was the last sample computed, get best times and update the ratio of ID to TLS.
+            if (Interlocked.Increment(ref sCalibrationDone) == CalibrationSamples)
+            {
+                double idMin = double.MaxValue;
+                double tlsMin = double.MaxValue;
+                for (int i = 0; i < CalibrationSamples; i++)
+                {
+                    idMin = Math.Min(idMin, calibrationState[i * 2]);       //ID
```

Comment:
Do we really need to keep the whole around to just compute the minimums at the end?